README.md
# Distributed Job Queue for Node.js backed by Redis

## How it works

 * Register tasks that can be performed. Each task definition lives in its
   own node module.
 * When you submit a processing job, you choose which task to run and
   supply a resource ID. If multiple processing jobs are submitted for the
   same resource ID, only one of them will run at a time.
 * Jobs are taken from a redis queue; so the task could be performed on any
   computer with access to the resource.
 * When you want to shutdown, call `shutdown` and wait for the callback. This
   will shutdown the queue gracefully, allowing any in progress jobs to
   complete. This works well in tandem with [naught](https://github.com/andrewrk/naught)
 * Workers can either run as child processes for extra concurrency and safety,
   or can run in the same memory context as your main process, for extra
   simplicity and convenience.
 * This module is only made to work with the Redis single instance configuration.
 * When a job fails or the process dies while a job is executing, the job is
   moved to a failed jobs queue. You can then retry failed jobs or delete them.

## Synopsis

```js
var JobQueue = require('redis-dist-job-queue');
var jobQueue = new JobQueue();

jobQueue.registerTask("./foo_node_module");

jobQueue.start();

var options = {resourceId: 'resource_id', params: {theString: "derp"}};
jobQueue.submitJob('thisIsMyTaskId', options, function(err) {
  if (err) throw err;
  console.info("job submitted");
});

jobQueue.on('error', function(err) {
  console.error("job queue error:", err.stack);
});
```

`foo_node_module.js`:

```js
module.exports = {
  id: 'thisIsMyTaskId',
  perform: function(params, callback) {
    console.info(params.theString.toUpperCase());
    callback();
  },
};
```


## Documentation

### JobQueue([options])

`options`:

 * `namespace` - redis key namespace. Defaults to `"redis-dist-job-queue."`
 * `queueId` - the ID of the queue on redis that we are connecting to.
   Defaults to `"default"`.
 * `redisConfig` - defaults to an object with properties:
   * `host` - 127.0.0.1
   * `port` - 6379
   * `db` - 0
 * `childProcessCount` - If set to 0 (the default), no child processes are
   created. Instead all jobs are performed in the current process. If set to
   a positive number, that many worker child processes are created.
 * `workerCount` - number of workers *for this JobQueue instance*, per child
   process. So if `childProcessCount` is positive, total concurrency for this
   server is `childProcessCount * workerCount`. If `childProcessCount` is `0`,
   total concurrency for this server is `workerCount`. Defaults to the number
   of CPU cores on the computer.
 * `flushStaleTimeout` - every this many milliseconds, scan for jobs that
   crashed while executing and moves them from the processing queue to the
   failed job queue. Defaults to 30000.

### jobQueue.start()

Starts the desired number of workers listening on the queue.

It's quite possible that you never want to call `start()`. For example, the
case where you want a server to submit processing jobs, but not perform them.

### jobQueue.registerTask(modulePath)

You must register all the tasks you want to be able to perform before calling
`start()`.

`modulePath` is resolved just like a call to `require`.

The node module at `modulePath` must export these options:

 * `id` - unique task ID that identifies what function to call to perform
   the task
 * `perform(params, callback)` - function which actually does the task.
   * `params` - the same params you gave to `submitJob`, JSON encoded
      and then decoded.
   * `callback(err)` - call it when you're done processing. Until you call
     this callback, the resource is locked and no other processing job will
     touch it.

And it may export these optional options:

 * `timeout` - milliseconds since last heartbeat to wait before considering
   a job failed. defaults to `10000`.

### jobQueue.submitJob(taskId, options, callback)

Adds a job to the queue to be processed by the next available worker.

 * `taskId` - the `id` field of a task you registered with `registerTask`
   earlier.
 * `options` - object containing any of these properties:
   * `resourceId` - a unique identifier for the resource you are about to
     process. Only one job will run at a time per resource.
   * `params` - an object which will get serialized to and from JSON and then
     passed to the `perform` function of a task.
   * `retries` - how many times to retry a job before moving it to the failed
     queue. Defaults to 0.
 * `callback(err)` - lets you know when the job made it onto the queue

### jobQueue.shutdown(callback)

Gracefully begins the shutdown process allowing any ongoing tasks to finish.
`callback` is called once everything is completly shut down.

### jobQueue.retryFailedJobs(callback)

Moves all jobs from the failed queue to the pending queue.

### jobQueue.deleteFailedJobs(callback)

Deletes all jobs from the failed queue.

### jobQueue.forceFlushStaleJobs(callback)

Forces an immediate flushing of jobs that crashed while executing. Jobs of this
sort are put into the failed queue, and you can then retry or delete them.
You probably don't want to call `forceFlushStaleJobs` manually. It's mostly for
testing purposes.


CHANGELOG.md
### 0.1.0

 * MIT LICENSE
 * update dependencies
 * use simpler and more robust uuid module
 * do not depend on hiredis
 * fix handling default config options

### 0.0.6

 * handle default config options better

### 0.0.5

 * add a `retries` parameter to `submitJob` which defaults to `0`.

### 0.0.4

 * when a job fails to acquire a lock, instead of dropping it, re-queue it

### 0.0.3

 * catch the case when the task returns a non error object
 * more consistent job.errorStack message
 * fix fake job failures from occuring


test/test.js
var JobQueue = require('..');
var assert = require('assert');
var path = require('path');
var shared = require('./shared');
var fs = require('fs');
var redis = require('redis');

var describe = global.describe;
var it = global.it;
var before = global.before;
var after = global.after;
var beforeEach = global.beforeEach;

var hitCountFile = path.resolve(__dirname, "hitCount.txt");
var twiceFile = path.resolve(__dirname, "twice.txt");

describe("JobQueue", function() {
  beforeEach(function(done) {
    var redisClient = redis.createClient();
    redisClient.send_command('flushdb', [], done);
  });

  after(function(done) {
    fs.unlink(hitCountFile, function(err) {
      done();
    });
  });

  before(tryUnlickTwiceFile);
  after(tryUnlickTwiceFile);

  function tryUnlickTwiceFile(done) {
    fs.unlink(twiceFile, function(err) {
      done();
    });
  }

  it("processes a thing and then shuts down - in process", function(done) {
    var jobQueue = new JobQueue({ workerCount: 8, redisConfig: {db: undefined} });
    shared.processedString = null;
    shared.callback = function(params) {
      assert.strictEqual(params, null);
      jobQueue.shutdown(function() {
        assert.strictEqual(shared.processedString, "DERP");
        done();
      });
    };

    jobQueue.registerTask('./task1');
    jobQueue.registerTask('./task_cb');

    jobQueue.start();

    jobQueue.submitJob('thisIsMyTaskId', {params: {theString: "derp"}}, assert.ifError);
    jobQueue.submitJob('callback', assert.ifError);
  });

  it("processes a thing and then shuts down - child processes", function(done) {
    var jobQueue = new JobQueue({ workerCount: 1, childProcessCount: 8 });

    fs.writeFileSync(hitCountFile, "0");

    jobQueue.registerTask('./task_hit_count');

    jobQueue.start();

    var options = {
      resourceId: 'hit_count',
      params: hitCountFile,
    };
    var count = 20;
    for (var i = 0; i < count; i += 1) {
      jobQueue.submitJob('hitCount', options, assert.ifError);
    }

    jobQueue.on('jobSuccess', function(err) {
      count -= 1;
      if (count === 0) return checkIt();
      if (count < 0) throw new Error("more job completions than expected");
    });

    function checkIt() {
      jobQueue.shutdown(function() {
        fs.readFile(hitCountFile, function(err, data) {
          if (err) throw err;
          assert.strictEqual(parseInt(data, 10), 20);
          done();
        });
      });
    }
  });

  it("moves jobs to failed queue and back", function(done) {
    shared.count = 0;

    var jobQueue = new JobQueue({ workerCount: 4 });

    jobQueue.once('jobFail', function() {
      jobQueue.retryFailedJobs(function(err) {
        if (err) return done(err);
      });
    });

    jobQueue.once('jobSuccess', function() {
      assert.strictEqual(shared.count, 2);
      jobQueue.shutdown(done);
    });

    jobQueue.registerTask('./task_fail');

    jobQueue.start();

    jobQueue.submitJob('failFirst', assert.ifError);
  });

  it("ability to delete failed jobs", function(done) {
    shared.count = 0;

    var jobQueue = new JobQueue({ workerCount: 1 });

    jobQueue.on('jobFail', function() {
      jobQueue.deleteFailedJobs(function(err) {
        if (err) return done(err);
        jobQueue.retryFailedJobs(function(err) {
          if (err) return done(err);
          shared.callback = function() {
            assert.strictEqual(shared.count, 1);
            jobQueue.shutdown(done);
          };
          jobQueue.submitJob('callback', assert.ifError);
        });
      });
    });

    jobQueue.registerTask('./task_fail');
    jobQueue.registerTask('./task_cb');

    jobQueue.start();

    jobQueue.submitJob('failFirst', assert.ifError);
  });

  it("handles jobs that crash while processing", function(done) {
    var jobQueue = new JobQueue({ childProcessCount: 3, workerCount: 1});

    jobQueue.once('childRestart', function() {
      jobQueue.once('jobFail', function(job) {
        assert.strictEqual(job.type, 'crashTask');
        jobQueue.shutdown(done);
      });

      setTimeout(function() {
        jobQueue.forceFlushStaleJobs(assert.ifError);
      }, 200);
    });


    jobQueue.registerTask('./task_crash');

    jobQueue.start();

    jobQueue.submitJob('crashTask', assert.ifError);
  });

  it("tries jobs twice if requested", function(done) {
    var jobQueue = new JobQueue();

    jobQueue.registerTask('./task_twice');

    jobQueue.on('jobSuccess', function(job) {
      jobQueue.shutdown(done);
    });

    jobQueue.start();

    jobQueue.submitJob('twice', {retries: 1}, assert.ifError);
  });

  it("retries jobs that time out", function(done) {
    var jobQueue = new JobQueue({childProcessCount: 1, workerCount: 1});

    jobQueue.registerTask('./task_twice_cb');

    jobQueue.on('jobSuccess', function(job) {
      jobQueue.shutdown(done);
    });

    jobQueue.start();

    jobQueue.submitJob('twice_cb', {params: twiceFile, retries: 1}, assert.ifError);

    setTimeout(function() {
      jobQueue.forceFlushStaleJobs(assert.ifError);
    }, 200);
  });
});


test/task1.js
var shared = require('./shared');

module.exports = {
  id: "thisIsMyTaskId",
  perform: function(params, callback) {
    shared.processedString = params.theString.toUpperCase();
    callback();
  },
};


test/task_cb.js
var shared = require('./shared');

module.exports = {
  id: "callback",
  perform: function(params, callback) {
    shared.callback(params);
    callback();
  },
};


lib/common.js
var redis = require('redis');
var crypto = require('crypto');

var htmlSafe = {'/': '_', '+': '-'};
var hasOwn = {}.hasOwnProperty;

exports.extend = extend;
exports.createRedisClient = createRedisClient;
exports.uuid = uuid;

function extend(target, source) {
  for (var propName in source) {
    if (hasOwn.call(source, propName)) {
      target[propName] = source[propName];
    }
  }
  return target;
}

function createRedisClient(config) {
  var client = redis.createClient(config.port, config.host, config);
  if (config.db != null) {
    client.select(config.db);
  }
  return client;
}

function uuid() {
  return rando(24).toString('base64').replace(/[\/\+]/g, function(x) {
    return htmlSafe[x];
  });
}

function rando(size) {
  try {
    return crypto.randomBytes(size);
  } catch (err) {
    return crypto.pseudoRandomBytes(size);
  }
}


lib/worker.js
var common = require('./common');
var createRedisClient = common.createRedisClient;
var extend = common.extend;
var Shavaluator = require('redis-evalsha');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var makeUuid = common.uuid;
var Pend = require('pend');

module.exports = Worker;

var shavaluator = new Shavaluator();

shavaluator.add('unlock',
    'if redis.call("get",KEYS[1]) == ARGV[1]\n' +
    'then\n' +
    '    redis.call("lrem",KEYS[2],1,ARGV[2])\n' +
    '    return redis.call("del",KEYS[1])\n' +
    'else\n' +
    '    return 0\n' +
    'end\n');

// KEYS[1] - the resource
// KEYS[2] - the pending jobs queue
// KEYS[3] - the processing jobs queue
// KEYS[4] - the failed jobs queue
// ARGV[1] - the job instance id
// ARGV[2] - the job payload
shavaluator.add('unlockMove',
    'if redis.call("get", KEYS[1]) == ARGV[1]\n' +
    'then\n' +
    '  local payload = ARGV[2]\n' +
    '  redis.call("lrem", KEYS[3], 1, payload)\n' +
    '  local json = cjson.decode(payload)\n' +
    '  local retries = json.retries or 0\n' +
    '  if retries > 0 then\n' +
    // we have more retries; put it back in the pending queue
    '    retries = retries - 1\n' +
    '    json.retries = retries\n' +
    '    payload = cjson.encode(json)\n' +
    '    redis.call("lpush", KEYS[2], payload)\n' +
    '  else\n' +
    // we have run out of retries; it goes into the failed jobs queue
    '    redis.call("lpush", KEYS[4], payload)\n' +
    '  end\n' +
    '  return redis.call("del",KEYS[1])\n' +
    'else\n' +
    '  return 0\n' +
    'end\n');

shavaluator.add('renew',
    'if redis.call("get",KEYS[1]) == ARGV[1]\n' +
    'then\n' +
    '    return redis.call("pexpire",KEYS[1],ARGV[2])\n' +
    'else\n' +
    '    return 0\n' +
    'end\n');

// KEYS[1] - source list
// KEYS[2] - dest list
// ARGV[1] - the list value to move
shavaluator.add('move',
    'redis.call("lrem",KEYS[1],1,ARGV[1])\n' +
    'redis.call("lpush",KEYS[2],ARGV[1])\n');

var taskDefaults = {
  // number of milliseconds since the last heartbeat to wait before
  // considering a job failed
  timeout: 10000,

  // perform: the actual task to do.
  perform: function(params, callback) {
    callback(new Error("implement this function"));
  },
};

util.inherits(Worker, EventEmitter);
function Worker(options) {
  EventEmitter.call(this);

  this.namespace = options.namespace;
  this.queueId = options.queueId;

  this.keyPrefix = this.namespace + this.queueId + ".";
  this.queueKey = this.keyPrefix + "queue";
  this.processingQueueKey = this.keyPrefix + "queue_processing";
  this.failedQueueKey = this.keyPrefix + "queue_failed";
  this.resourceKeyPrefix = this.keyPrefix + "lock.";
  this.workerCount = options.workerCount;
  this.redisConfig = options.redisConfig;
  this.tasks = {};
  
  for (var i = 0; i < options.modulePaths.length; i += 1) {
    var taskModule = require(options.modulePaths[i]);
    this.tasks[taskModule.id] = extend(extend({}, taskDefaults), taskModule);
  }
}

Worker.prototype.start = function() {
  this.shuttingDown = false;
  this.pend = new Pend();
  this.blockingRedisClients = [];
  this.redisClientStatus = [];
  for (var i = 0; i < this.workerCount; i += 1) {
    this.spawnWorker();
  }
}

Worker.prototype.shutdown = function(callback) {
  var self = this;
  self.shuttingDown = true;
  self.blockingRedisClients.forEach(function (info) {
    if (info.blocking) {
      info.redisClient.end();
    }
  });
  self.pend.wait(function() {
    callback();
  });
};

Worker.prototype.spawnWorker = function() {
  var self = this;

  // create new client because blpop blocks the connection
  var redisClient = createRedisClient(self.redisConfig);
  var clientObj = {
    redisClient: redisClient,
    blocking: false,
  };
  self.blockingRedisClients.push(clientObj);

  handleOne();

  function handleOne() {
    if (self.shuttingDown) return;
    self.processOneItem(clientObj, handleOne);
  }
};

Worker.prototype.processOneItem = function(clientObj, cb) {
  var self = this;

  var redisClient = clientObj.redisClient;
  clientObj.blocking = true;
  redisClient.send_command('brpoplpush', [self.queueKey, self.processingQueueKey, 0], onResult);

  function onResult(err, jobListValue) {
    // so later we can do pend.wait and get a callback when
    // shutting down
    clientObj.blocking = false;
    self.pend.go(function(pendCb) {
      goForIt(err, jobListValue, function() {
        pendCb();
        cb();
      });
    });
  }

  function goForIt(err, jobListValue, cb) {
    if (err) {
      self.emit('error', err);
      return cb();
    }

    var job;
    try {
      job = JSON.parse(jobListValue);
    } catch (err) {
      self.emit('error', err);
      return cb();
    }

    var task = self.tasks[job.type];
    if (!task) {
      self.emit('error', new Error("unregistered task: " + job.type));
      return cb();
    }

    var heartbeatTimeout = Math.floor(task.timeout / 2);
    if (isNaN(heartbeatTimeout)) {
      self.emit('error', new Error("invalid heartbeat timeout"));
      return cb();
    }

    var resourceKey = self.resourceKeyPrefix + job.resource;
    var jobInstanceId = makeUuid();
    var args = [resourceKey, jobInstanceId, 'NX', 'PX', task.timeout];
    redisClient.send_command('set', args, onSetResult);
    
    function onSetResult(err, result) {
      if (err) {
        self.emit('error', err);
        return cb();
      }

      var interval = null;

      // If we fail to acquire the lock, it means that this resource is
      // already being processed. We move the item from the processing
      // queue to the end of the waiting queue and then GTFO
      if (!result) {
        shavaluator.execWithClient(redisClient, 'move',
            [self.processingQueueKey, self.queueKey], [jobListValue],
            function(err, result)
        {
          if (err) {
            self.emit('error', err);
          }
          cb();
        });
        return;
      }

      // start the heart beating. This prevents the job from timing out
      // while it is still running.
      interval = setInterval(doHeartBeat, heartbeatTimeout);

      // emit an error if the user calls the callback twice
      var callbackCalled = false;

      // now actually do the thing
      task.perform(job.params, onComplete);

      function doHeartBeat() {
        shavaluator.execWithClient(redisClient, 'renew',
          [resourceKey], [jobInstanceId, task.timeout], function(err, result)
        {
          if (err) {
            self.emit('error', err);
          }
          if (err || !result) {
            if (interval) clearInterval(interval);
            interval = null;
          }
        });
      }

      function onComplete(err) {
        if (callbackCalled) {
          self.emit('error', new Error(job.type + ": callback called twice"));
          return;
        }
        callbackCalled = true;

        if (interval) clearInterval(interval);
        if (err) {
          if (!(err instanceof Error)) {
            err = new Error("task returned non error object");
          }
          job.errorStack = err.stack;
          shavaluator.execWithClient(redisClient, 'unlockMove',
            [resourceKey, self.queueKey, self.processingQueueKey, self.failedQueueKey],
            [jobInstanceId, jobListValue], finishFail);
        } else {
          shavaluator.execWithClient(redisClient, 'unlock',
            [resourceKey, self.processingQueueKey],
            [jobInstanceId, jobListValue], finishSuccess);
        }

        function finishFail(err) {
          self.emit('jobFail', job);
          if (err) {
            self.emit('error', err);
          }
          cb();
        }

        function finishSuccess(err) {
          self.emit('jobSuccess', job);
          if (err) {
            self.emit('error', err);
          }
          cb();
        }
      }
    }
  }
};


lib/index.js
var common = require('./common');
var createRedisClient = common.createRedisClient;
var extend = common.extend;
var redis = require('redis');
var Shavaluator = require('redis-evalsha');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var makeUuid = common.uuid;
var cpuCount = require('os').cpus().length;
var Pend = require('pend');
var path = require('path');
var JobWorker = require('./worker');
var spawn = require('child_process').spawn;

module.exports = JobQueue;

var childModulePath = path.join(__dirname, "child.js");

var queueDefaults = {
  namespace: "redis-dist-job-queue.",
  queueId: "default",
  redisConfig: {},
  workerCount: cpuCount,
  childProcessCount: 0,
  flushStaleTimeout: 30000,
};

var redisConfigDefaults = {
  port: 6379,
  host: "127.0.0.1",
  db: 0,
};

util.inherits(JobQueue, EventEmitter);
function JobQueue(options) {
  EventEmitter.call(this);

  options = extend(extend({}, queueDefaults), options || {});
  this.namespace = options.namespace;
  this.queueId = options.queueId;

  this.keyPrefix = this.namespace + this.queueId + ".";
  this.queueKey = this.keyPrefix + "queue";
  this.processingQueueKey = this.keyPrefix + "queue_processing";
  this.scratchQueueKey = this.keyPrefix + "queue_scratch";
  this.failedQueueKey = this.keyPrefix + "queue_failed";
  this.resourceKeyPrefix = this.keyPrefix + "lock.";
  this.flushStaleKey = this.keyPrefix + "flushStaleLock";
  this.flushStaleTimeout = options.flushStaleTimeout;
  this.workerCount = options.workerCount;
  this.childProcessCount = options.childProcessCount;
  this.redisConfig = extend(extend({}, redisConfigDefaults), options.redisConfig);
  this.modulePaths = [];
  this.redisClient = createRedisClient(this.redisConfig);
  this.childProcesses = [];
  this.childWorker = null;

  this.shavaluator = new Shavaluator();

  // KEYS[1] - pending job queue list
  // KEYS[2] - processing list
  // KEYS[3] - failed jobs list
  // KEYS[4] - scratch space list
  // KEYS[5] - flush stale timeout resource lock
  // ARGV[1] - the string to use for the flush stale timeout resource lock
  // ARGV[2] - number of milliseconds to hold the lock
  this.shavaluator.add('flushStale',
    // move list items one by one from source list to either scratch space
    // list or dest list.
    'local newFailedJobs = {}\n' +
    'if not redis.call("set", KEYS[5], ARGV[1], "NX", "PX", ARGV[2]) then\n' +
    '  return newFailedJobs\n' +
    'end\n' +
    'while true do\n' +
    '  local payload = redis.call("rpop", KEYS[2])\n' +
    '  if not payload then\n' +
    '    break\n' +
    '  end\n' +
    '  local json = cjson.decode(payload)\n' +
    '  local resourceId = json.resource\n' +
    '  local key = "' + this.resourceKeyPrefix + '" .. resourceId\n' +
    '  local ttl = redis.call("pttl", key)\n' +
    '  if ttl < 0 then\n' +
    '    local retries = json.retries or 0\n' +
    '    if retries > 0 then\n' +
    // we have retries left; move to end of pending queue
    '      retries = retries - 1\n' +
    '      json.retries = retries\n' +
    '      payload = cjson.encode(json)\n' +
    '      redis.call("lpush", KEYS[1], payload)\n' +
    '    else\n' +
    // ran out of retries, task is moved to failed list
    '      redis.call("lpush", KEYS[3], payload)\n' +
    '      table.insert(newFailedJobs, payload)\n' +
    '    end\n' +
    '  else\n' +
    '    redis.call("lpush", KEYS[4], payload)\n' +
    '  end\n' +
    'end\n' +
    // now we rename the scratch space list to the processing list
    'redis.pcall("rename", KEYS[4], KEYS[2])\n' +
    'return newFailedJobs\n');

  // KEYS[1] - source list
  // KEYS[2] - dest list
  this.shavaluator.add('moveAll',
      'while redis.call("rpoplpush",KEYS[1],KEYS[2]) do\n' +
      'end\n' +
      'return nil\n');
}

JobQueue.prototype.start = function() {
  var self = this;
  self.shuttingDown = false;
  if (self.childProcessCount > 0) {
    self.startChildProcesses();
  } else {
    self.startWorkers();
  }
  self.flushStaleInterval = setInterval(doFlushStaleJobs, self.flushStaleTimeout);
  doFlushStaleJobs();

  function doFlushStaleJobs() {
    self.flushStaleJobs(function(err) {
      if (err) {
        self.emit('error', err);
      }
    });
  }
};

JobQueue.prototype.forceFlushStaleJobs = function(callback) {
  var self = this;
  self.redisClient.send_command('del', [self.flushStaleKey], function(err) {
    if (err) return callback(err);
    self.flushStaleJobs(callback);
  });
};

JobQueue.prototype.flushStaleJobs = function(callback) {
  var self = this;
  self.shavaluator.execWithClient(self.redisClient, 'flushStale',
      [self.queueKey, self.processingQueueKey, self.failedQueueKey, self.scratchQueueKey,
      self.flushStaleKey], ["1", self.flushStaleTimeout],
      function(err, failedJobs)
  {
    if (err) {
      var newErr = new Error("flushStaleJobs failed");
      newErr.stack += "\n" + err.stack;
      return callback(newErr);
    }
    failedJobs.forEach(function(failedJob) {
      var job = JSON.parse(failedJob);
      job.errorStack = new Error("timed out").stack;
      self.emit('jobFail', job);
    });
    callback();
  });
};

JobQueue.prototype.startChildProcesses = function() {
  var self = this;
  var args = [
    childModulePath,
    JSON.stringify(self.serializeOptions()),
  ];
  var opts = {
    stdio: [process.stdin, process.stdout, process.stderr, 'ipc'],
  };
  for (var i = 0; i < self.childProcessCount; i += 1) {
    createChild();
  }

  function createChild() {
    var child = spawn(process.execPath, args, opts);
    self.childProcesses.push(child);
    child.on('exit', createOnExit(child));
    child.on('message', onMessage);
  }

  function createOnExit(child) {
    return function() {
      var index = self.childProcesses.indexOf(child);
      if (index >= 0) self.childProcesses.splice(index, 1);
      if (self.shuttingDown) return;
      self.emit('childRestart');
      createChild();
    };
  }

  function onMessage(msg) {
    if (self.shuttingDown) return;

    if (msg.type === 'error') {
      self.emit('error', new Error(msg.value));
    } else if (msg.type === 'jobFail') {
      self.emit('jobFail', msg.value);
    } else if (msg.type === 'jobSuccess') {
      self.emit('jobSuccess', msg.value);
    } else {
      throw new Error("unrecognized message type: " + msg.type);
    }
  }
};

JobQueue.prototype.serializeOptions = function() {
  return {
    namespace: this.namespace,
    queueId: this.queueId,
    workerCount: this.workerCount,
    redisConfig: this.redisConfig,
    modulePaths: this.modulePaths,
  };
};

JobQueue.prototype.startWorkers = function() {
  var self = this;
  self.childWorker = new JobWorker(self.serializeOptions());
  self.childWorker.on('error', function(err) {
    self.emit('error', err);
  });
  self.childWorker.on('jobFail', function(job) {
    self.emit('jobFail', job);
  });
  self.childWorker.on('jobSuccess', function(job) {
    self.emit('jobSuccess', job);
  });
  self.childWorker.start();
};

JobQueue.prototype.registerTask = function(modulePath) {
  var dirname = path.dirname(module.parent.filename);
  var modulePathAbs = require.resolve(path.resolve(dirname, modulePath));
  this.modulePaths.push(modulePathAbs);
};

// begin a process job. If the resource is already ongoing processing,
// nothing happens.
JobQueue.prototype.submitJob = function(taskId, options, cb) {
  options = options || {};
  if (typeof options === 'function') {
    cb = options;
    options = {};
  }

  var resourceId = options.resourceId || makeUuid();
  var params = options.params || null;
  var retries = parseInt(options.retries, 10) || 0;
  var json = JSON.stringify({
    type: taskId,
    resource: resourceId,
    params: params,
    retries: retries,
  });
  this.redisClient.send_command('lpush', [this.queueKey, json], function(err, result) {
    cb(err);
  });
};

JobQueue.prototype.shutdown = function(callback) {
  var self = this;

  self.shuttingDown = true;
  clearInterval(self.flushStaleInterval);

  var pend = new Pend();
  pend.go(function(cb) {
    self.redisClient.quit(function(err, result) {
      if (err) {
        self.emit('error', err);
      }
      cb();
    });
  });
  self.childProcesses.forEach(function(child) {
    pend.go(function(cb) {
      child.on('exit', cb);
      child.send('shutdown');
    });
  });
  if (self.childWorker) {
    pend.go(shutdownChildWorker);
  }
  pend.wait(function() {
    callback();
  });

  function shutdownChildWorker(cb) {
    self.childWorker.shutdown(cb);
  }
};

JobQueue.prototype.deleteFailedJobs = function(callback) {
  this.redisClient.send_command('del', [this.failedQueueKey], callback);
};

JobQueue.prototype.retryFailedJobs = function(callback) {
  this.shavaluator.execWithClient(this.redisClient, 'moveAll',
      [this.failedQueueKey, this.queueKey], [], callback);
}
